### ISSUE TITLE
The model structure fracture occurs in freezing training

### ISSUE THREAD

**Issue by Dasenyoung** [ultralytics/ultralytics#20134]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have 4 types of data, among which two types of detection effect is very poor (the two types are very similar and both are small targets, and one type of features is very difficult and not obvious), with about 75% accuracy. yolov11 is used for training detection, and now a model has been trained. In order to solve the problem of those two types of effects, I made some data sets containing only these two types, loaded the trained model, froze the remaining layers, only kept the detection head part, and reduced the learning rate and rounds, but the training log showed that the effect was only 3%. I thought it was because I only trained the detection head that it was low and I wasn't worried. But the resulting model I used was still only 3% accurate? Although the rest of the layer is not trained, but the ability is still ah, why the final accuracy is as low as the training log? And the rest of the floors weren't damaged? The feeling is that the model architecture is broken, and the entire yolo is left with one detection head, and the rest is removed. Here is my yaml with training code:
## YOLOv12 object detection model with P3-P5 outputs. For Usage examples see https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 4 # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov12n.yaml' will call yolov12.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.50, 0.25, 1024] # summary: 465 layers, 2,603,056 parameters, 2,603,040 gradients, 6.7 GFLOPs
  s: [0.50, 0.50, 1024] # summary: 465 layers, 9,285,632 parameters, 9,285,616 gradients, 21.7 GFLOPs
  m: [0.50, 1.00, 512] # summary: 501 layers, 20,201,216 parameters, 20,201,200 gradients, 68.1 GFLOPs
  l: [1.00, 1.00, 512] # summary: 831 layers, 26,454,880 parameters, 26,454,864 gradients, 89.7 GFLOPs
  x: [1.00, 1.50, 512] # summary: 831 layers, 59,216,928 parameters, 59,216,912 gradients, 200.3 GFLOPs

# YOLO12n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv,  [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv,  [128, 3, 2]] # 1-P2/4
  - [-1, 2, C3k2,  [256, False, 0.25]]# 2
  - [-1, 1, Conv,  [256, 3, 2]] # 3-P3/8
  - [-1, 2, C3k2,  [512, False, 0.25]]# 4
  - [-1, 2, FEM, []]# 5
  - [4, 1, Conv,  [512, 3, 2]] # 6-P4/16
  - [-1, 4, A2C2f, [512, True, 4]]# 7
  - [-1, 2, FEM, []]# 8
  - [7, 1, Conv,  [1024, 3, 2]] # 9-P5/32
  - [-1, 4, A2C2f, [1024, True, 1]] # 10
  - [-1, 2, FEM, []]# 11

# YOLO12n head
head:
  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]# 12
  - [[-1, 8], 1, Concat, [1]] # cat backbone P4 #13
  - [-1, 2, A2C2f, [512, False, -1]] # 14

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]# 15
  - [[-1, 5], 1, Concat, [1]] # cat backbone P3#16
  - [-1, 2, A2C2f, [256, False, -1]] # 17

  - [-1, 1, Conv, [256, 3, 2]]#18
  - [[-1, 14], 1, Concat, [1]] # cat head P4#19
  - [-1, 2, A2C2f, [512, False, -1]] # 20

  - [-1, 1, Conv, [512, 3, 2]]#21
  - [[-1, 11], 1, Concat, [1]] # cat head P5#22
  - [-1, 2, C3k2, [1024, True]] # 23 (P5/32-large)

  - [ 17, 1, SCAM, []] # 24
  - [ 20, 1, SCAM, []] # 25
  - [ 24, 1, SCAM, []] # 26

  - [[24, 25, 26], 1, Detect, [nc]] # Detect(P3, P4, P5)

from ultralytics.models import YOLO
import os

os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'

if __name__ == '__main__':
    model = YOLO(model='ultralytics/cfg/models/v12/yolov12s-FEM-SCAM.yaml')
    model.load('./runs-you1200s-FEM-SCAM/train\exp\weights/best.pt')
    model.train(data='./data1.yaml', epochs=150, batch=32, device='0', imgsz=640, freeze=24, lr0=0.001, scale=0.9,  mosaic=1.0, workers=32, cache=False, mixup=0.2, copy_paste=0.6,
                amp=False, project='runs-you1200s-FEM-SCAM-Fine-tuning/train', name='exp')

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Dasenyoung, thank you for reaching out and sharing your detailed query with us 🚀! We appreciate your interest in YOLO12 and your effort in providing your training setup and YAML configuration.

If this is a 🐛 Bug Report, we kindly ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug the issue effectively. This would include the dataset or a subset of it, along with a simplified script reproducing the problem.

If this is a custom training ❓ Question, please ensure you’ve included:
- Dataset details or examples (e.g., representative images).
- Training logs showing the observed behavior.
- Confirmation that you’ve followed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Additionally, please upgrade to the latest version of the `ultralytics` package to ensure you are on the most up-to-date release that might already address this issue. You can upgrade by running:

```bash
pip install -U ultralytics
```

We also recommend testing your setup in one of YOLO's verified environments to ensure all dependencies are correctly configured:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For further exploration, you can find helpful information in the [Ultralytics Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.

Lastly, feel free to join our growing Ultralytics community for discussions and support:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) for knowledge sharing

An Ultralytics engineer will review your issue soon and assist further. Thank you for your patience and for using Ultralytics YOLO 😊!

**Comment by glenn-jocher** [comment]
Looking at your training setup, I think I see what's happening. When you freeze layers using `freeze=24`, you're freezing everything up to the SCAM module in your architecture, but the issue may be in how you're fine-tuning with a subset of your original classes.

Since you're training on a dataset that only contains 2 of the 4 original classes, but your model is still configured for 4 classes (`nc: 4`), there's likely a class mismatch. The model expects 4 classes during training, but your fine-tuning dataset only provides 2 classes.

Try these approaches:
1. Make sure your fine-tuning data.yaml has the same class structure as your original training (all 4 classes), even if you're only including images from 2 classes
2. Lower your freeze level to allow more of the model to adapt (try `freeze=20` or even lower)
3. Try using a higher learning rate for fine-tuning (0.001 may be too low for the head to adapt properly)
4. Consider training for longer with early stopping by setting `patience=50`

Let me know if these suggestions help improve your results.

**Comment by Y-T-G** [comment]
It doesn't seem like you're using official Ultralytics either. The code you posted is for the YOLO12 fork, not Ultralytics. You should use official Ultralytics since this issues section is for help with official Ultralytics, not unofficial forks. For issues with the fork, you can open an issue on their respective repositories.

**Comment by Dasenyoung** [comment]
> Looking at your training setup, I think I see what's happening. When you freeze layers using `freeze=24`, you're freezing everything up to the SCAM module in your architecture, but the issue may be in how you're fine-tuning with a subset of your original classes.
> 
> Since you're training on a dataset that only contains 2 of the 4 original classes, but your model is still configured for 4 classes (`nc: 4`), there's likely a class mismatch. The model expects 4 classes during training, but your fine-tuning dataset only provides 2 classes.
> 
> Try these approaches:
> 
> 1. Make sure your fine-tuning data.yaml has the same class structure as your original training (all 4 classes), even if you're only including images from 2 classes
> 2. Lower your freeze level to allow more of the model to adapt (try `freeze=20` or even lower)
> 3. Try using a higher learning rate for fine-tuning (0.001 may be too low for the head to adapt properly)
> 4. Consider training for longer with early stopping by setting `patience=50`
> 
> Let me know if these suggestions help improve your results.

I understand what you mean. Even if I load a model trained on 4 categories, no matter how many layers I freeze the front part of the model, the new dataset only has 2 categories, even if these 2 categories belong to the 4 categories, then the trained new model will only be able to detect and recognize 2 categories, right? So if I only freeze the detection head part (how to achieve it?), or fine-tune the backbone and neck with the 2-category new dataset, will this enhance the backbone's ability to detect these 2 categories while also maintaining the ability to detect and recognize 4 categories? Or will this only make the backbone only have the ability to extract features of these 2 categories of targets, and the other 2 categories will not work?

**Comment by glenn-jocher** [comment]
You've asked really important questions about fine-tuning. Let me clarify:

When fine-tuning with a subset of classes, freezing works differently depending on which parts you freeze:

1. To freeze only the backbone and neck (allowing the detection head to train): Try using a negative freeze value like `freeze=-3` which will freeze all layers except the last 3 (typically the detection head). This approach preserves feature extraction while allowing the head to adapt.

2. For your specific question: If you fine-tune on only 2 classes but keep your model configured for 4 classes (in both YAML and data.yaml), and primarily train the head, you can maintain detection capabilities for all 4 classes while improving performance on the 2 problematic ones.

3. The backbone primarily learns general feature extraction capabilities. Fine-tuning it on 2 classes won't make it "forget" features for the other classes, especially if you use a lower learning rate. However, if you completely retrain the backbone with only 2 classes at a high learning rate, it might become biased toward those classes.

I'd recommend creating a dataset that includes all 4 classes but oversamples the 2 problematic ones, then fine-tune with a relatively low learning rate (0.0001-0.0005) to improve detection while maintaining overall performance.

**Comment by Dasenyoung** [comment]
> You've asked really important questions about fine-tuning. Let me clarify:
> 
> When fine-tuning with a subset of classes, freezing works differently depending on which parts you freeze:
> 
> 1. To freeze only the backbone and neck (allowing the detection head to train): Try using a negative freeze value like  which will freeze all layers except the last 3 (typically the detection head). This approach preserves feature extraction while allowing the head to adapt.`freeze=-3`
> 2. For your specific question: If you fine-tune on only 2 classes but keep your model configured for 4 classes (in both YAML and data.yaml), and primarily train the head, you can maintain detection capabilities for all 4 classes while improving performance on the 2 problematic ones.
> 3. The backbone primarily learns general feature extraction capabilities. Fine-tuning it on 2 classes won't make it "forget" features for the other classes, especially if you use a lower learning rate. However, if you completely retrain the backbone with only 2 classes at a high learning rate, it might become biased toward those classes.
> 
> I'd recommend creating a dataset that includes all 4 classes but oversamples the 2 problematic ones, then fine-tune with a relatively low learning rate (0.0001-0.0005) to improve detection while maintaining overall performance.

Thank you for your reply! I used a mix of 4 types of data (oversampled 2 of them), loaded the original pre-trained model, and the fine-tuning training got good results! Now I have understood the truth about fine tuning, this post is finished!
